### PR TITLE
Adding wrapper functions for advance, predict and solve operations.

### PR DESCRIPTION
**Pull request by jcotela**
@philbucher detected that, in the merged design of the analysis/stage class, we did not provide wrapper functions for the `AdvanceInTime`, `Predict` and `SolveSolutionStep` methods of the solver.

In the case of black-box coupling, this would force the user of the analysis class to know its implementation details (the underlying solver object) in order to implement the coupled simulation loop. We both believe that this would be a bad design, so we propose adding the functions included in this PR. We will update the wiki accordingly if this is approved.

Update: also changed the name of the file to match the name of the class

### CODE REVIEW

File: kratos/python_scripts/analysis_stage.py [hunk 1]
```diff
@@ -49,15 +49,13 @@ def RunSolutionLoop(self):
...
         while self.time < self.end_time:
-            solver.AdvanceInTime(self.time)
+            self.AdvanceInTime()
             self.InitializeSolutionStep()
-            solver.Predict()
-            solver.SolveSolutionStep()
+            self.Predict()
+            self.SolveSolutionStep()
             self.FinalizeSolutionStep()
             self.OutputSolutionStep()
```
Comment (adityaghantasala) [inline kratos/python_scripts/analysis_stage.py:55]: Isn't there ExecuteAfterSolutionLoop for all the processes missing here ?

```
for process in self.list_of_processes:
     process.ExecuteAfterSolutionLoop()
```

Comment (jcotela) [reply kratos/python_scripts/analysis_stage.py:55]: I had to check, but apparently the `process.ExecuteAfterSolutionLoop()` does not exist (there is only Finalize)

**Comment by RiccardoRossi** [comment]
@jcotela i would say that the choice of using "solver." instead of "self." was deliberate, since one is supposed to write a solver for that...

having said this i am personally not against what you propose

**Comment by RiccardoRossi** [comment]
expanding on my previous answer:

my call is to create a BlackBoxCouplingSolver, and to use together with the standard stage. So we avoid proliferation of stages...

**Comment by maceligueta** [comment]
@RiccardoRossi , at the @KratosMultiphysics/technical-committee  we agreed that we would leave the door open to both types of couplings: coupling stages and BlackBoxCouplingSolver's. So the developers would choose the style they prefer, right?
@jcotela Sorry, I don't quite understand what you mean as black-box coupling. Actually I am in favour of the changes in this PR, but I would like to understand the problem you and @philbucher found.

**Comment by jcotela** [comment]
@maceligueta The idea is to have a coupling stage (we are thinking FSI but the same idea would apply to anything else) that works by combining the functions in the existing standard stages. This stage would have as members instances of the standard fluid and structure stages and call their methods in the right order, defining a coupling loop in the `AnalysisStage.RunSolutionLoop()` function.

In this approach, the proposed new functions would allow us to achieve this without having to call the solver for each stage, which is an implementation detail from the point of view of the coupled stage. I was under the impression that having this possibility was the rationale for dividing the main stage solution loop in individual functions, so here we were trying to provide the "missing pieces".

**Comment by RiccardoRossi** [comment]
it is my understanding that it is ALLOWED to write your own stage. The preferred choice is however to write a new solver.

For this reason my understanding is that 

       solver.

was used instead of 

      self.

exactly since someone that has to modify that function needs some knowledte of the internals.

My call is still that the best solution would be to implement a coupling_solver and use it within a "normal" dynamic stage.
The main point here is that you can compose different solvers but **you cannot compose different stages**. Think for example you have a fsi_solver and a thermal_solver, if they are both solvers you can combine them in a fsi_thermal_solver (you may even use a black_box_coupling_solver nested within another black_box_coupling_solver!!). If they are stages you would need to write a new stage for that.

**Comment by philbucher** [comment]
@RiccardoRossi I am not sure I can follow you, so here is a proposal:

I spoke a lot with @jcotela and we discussed the details of how the AnalysisStage should be implemented for Fluid and Structure, in fact we started with this already.

Since you are not against the proposed changes and also @maceligueta is ok with it I would ask you to approve.

We can then finish the implementation and show you examples of how it is being used e.g. in FSI
I think it will be a lot easier to discuss on actual examples

Otherwise I fear that this fill be the next 50+ comments thread and I feel like we are turning in circles for a while already.

**Comment by RiccardoRossi** [comment]
I am sorry @philbucher but i am on disagreement on this. 

I think [and i appreciate we may disagree about this] that writing a FSISolver is a superior solution, and i think it would be the way to go here. For this reason i see no reason for which we should change the proposed design. 

I think that my argument of combining a FSISolver + thermal-solver is a good one, and i don't see how you would do that with a stage.

Having said this i am NOT using my review to block on this. To my understanding you already have 2 member of the @KratosMultiphysics/technical-committee in agreement with you, so if there is a third one in agreement you can proceed. I'll live with it :-)

**Comment by jcotela** [comment]
@RiccardoRossi sorry but I also don't follow. Why can you compose solvers but not compose stages?

When this was debated by the @KratosMultiphysics/technical-committee, I also got the impression that the door was open to coupling both at the stage level (if the two solvers can work independently without knowing each other's internals) and at the solver level (if this is not the case, as in the thermo-fluid solver with coupled nodes).

Anyways, I don't think this is something that needs to be decided by a majority vote. Instead, I'd like to understand the reasoning for your position.

**Comment by philbucher** [comment]
I fully agree with @jcotela 

Another thought:
regardless of the coupling even for pure cases (e.g. in a standard structural analysis if you want to print some values after the `Predict` function) the AnalysisStage should expose the functions that are added in this PR.

**Comment by RiccardoRossi** [comment]
Hi guys, 
         sorry for delay in answering. I was convinced that i did answer yesterday but apparently i made smthg wrong.

Let's consider a simple FSI solver. Solve function looks smthg lik

       fsi.Solve()
            while(not converged):
                 structure.Solve()
                 map_str_to_fluid
                 fluid.Solve()
                 map_fl_to_str

If we want to solve a normal fluid + thermal we would do
 
       fluid.Solve()
       th.Solve()

which can be easily imlemeneted in a "general purpose solver as"

       solver1.Solve()
       solver2.Solve()

where solver1=fluid, solver2 = structure.

Now my point is that changing such solver to make fsi instead of fluid alone is simply telling that solver1=fsi.

How would you achieve the same if you have an fsi stage instead of an fsi solver?

**Comment by philbucher** [comment]
Hi @RiccardoRossi 
In my eyes the big advantage of having an FSI stage over having an FSI solver is not having to know abt the processes, correct order of Initialization etc any more.
Previously whenever we did FSI we had to check if sth had changed and figure it out ourselves. This would not be necessary any more if we use the stage, which contains ALL the information.

An FSI stage would therefore look as simple as this:
~~~py

import fluid_stage
import structure_stage

class FSIAnalysisStage(BaseAnalysisStage):

    def __init__(self):
        self.fluid = FluidStage()
        self.structure = StructureStage()

    def Initialize(self):
        self.fluid.Initialize()
        self.structure.Initialize()

    def InitializeSolutionStep(self):
        self.fluid.InitializeSolutionStep()
        self.structure.InitializeSolutionStep()

    def Predict(self):
        self.fluid.Predict()
        self.structure.Predict()

    def SolveSolutionStep(self):
        while(not converged):
            self.fluid.SolveSolutionStep()
            self.MapForces()
            self.Structure.SolveSolutionStep()
            self.MapDisplacements()

    def FinalizeSolutionStep():
        fluid.FinalizeSolutionStep()
        structure.FinalizeSolutionStep()

    def Finalize(self):
        self.fluid.Finalize()
        self.structure.Finalize()
~~~
Even though this is just written down briefly this is actually not far from how it would look like I think

If you want thermal+fluid instead of just fluid this would actually be implemented as a new solver (I think this is done already, same for ALE+fluid) since this is a more "tight" coupling due to shared nodes etc

**Comment by RiccardoRossi** [comment]
@philbucher , if you replace the name "stage" by "solver" in what you wrote you can obtain the same result + the possibility to compose it with the thermal solver ...

**Comment by jcotela** [comment]
@RiccardoRossi where will you put the coupled "FSI + thermal" solver? I guess you will anyway need to write a new solver (or a new stage) for that. In the end, the amount of code is the same, unless I am missing something.

**Comment by philbucher** [comment]
I moved the `ExecuteBeforeSolutionLoop` calls to the processes to `Initialize`, because otherwise I cannot use the Stage without calling `RunSolutionLoop`

**Comment by philbucher** [comment]
Another non-negligible disadvantage of not allowing Stage-Coupling is that you have to have all ProjectParameters in ONE file

**Comment by RiccardoRossi** [comment]
My proposal for the thermal + fsi would be the following

first let's define a "generic" composite solver, which simply calls two solvers one after the other

    class CompositeSolver(BaseAnalysisStage, settings):
           def __init__(self, settings):
                solver1 ... construct with settings["solver1"]
                solver2 ... construct with settings["solver2"]

          ... here all functions needed calling solver1 & solver2. for example:

          def Predict(self):
                  self.solver1.Predict()
                  self.solver2.Predict()

          def SolveSolutionStep(self):
                  self.solver1.SolveSolutionStep()
                  self.solver2.SolveSolutionStep()

          ...

then we define the "FSI" solver, as complex as we would like it to be. note that here i am using copy pasted your design and i simply called it a solver

    class FSISolver(BaseAnalysisStage):

        def __init__(self, settings):

            
            self.fluid = FluidSolver()
            self.structure = StructureSolver()

        def Initialize(self):
            self.fluid.Initialize()
            self.structure.Initialize()

        def InitializeSolutionStep(self):
            self.fluid.InitializeSolutionStep()
            self.structure.InitializeSolutionStep()

        def Predict(self):
            self.fluid.Predict()
            self.structure.Predict()

        def SolveSolutionStep(self):
            while(not converged):
                self.fluid.SolveSolutionStep()
                self.MapForces()
                self.Structure.SolveSolutionStep()
                self.MapDisplacements()

        def FinalizeSolutionStep():
            fluid.FinalizeSolutionStep()
            structure.FinalizeSolutionStep()

        def Finalize(self):
            self.fluid.Finalize()
            self.structure.Finalize()

finally we call the "BaseAnalysisStage" passing a project parameters file that simply nests the solver settings, within a "solver1" tag and "solver2" tag of the composite solver. @philbucher this means that everything can be in a single file, just nested

     {
            ... here settings for the stage, which are "standard" since we use the BaseStage.
            solver_settings : {
                  "solver_type" : "CompositeSolver",
                   "solver1_settings" : { ... here the block witht the FSI settings }
                   "solver2_settings" : { ... here the block witht the thermal settings }
                   ....
                  }
           ...here processes for the BCs ...
     }

As you see the design is pretty similar. The difference is that (in my opinion) putting everything in the stage violates two decisions (however stage avoids composition- you need to reimplement the runsolutionloop!!):

1. it exposes to the user "inner details" of the fsi solver: the mapping for example (note that you are already following the design i propose for the AleSolver)
2. you are putting knowledge of the phsyics into the stage layer, which by design does not know it

**Comment by philbucher** [comment]
@RiccardoRossi one question remains for me: when you speak about solver, do you mean the ones that are currently in Fluid and Structural?

If so, then my concern is that I will have to manually compose the processes etc which is error prone (this is what we do now and we have to update it all the time)

Can the "Solver" be used standalone (i.e. it sets up its processes etc) to run a simulation?

**Comment by RiccardoRossi** [comment]
@philbucher I am not sure if i understand your concern. Let me have a try to see if ido: I think we need to distinguish something here.

The "solver" calls internally "utilities" or eventually the "process.Execute()" function in the case processes are used internally. **Users should not be aware of those**.

the processes for the application of BCs should be called only from the stage level, and should be provided by the user through the ProjectParameters.json. So no composition should be needed here...

**Comment by philbucher** [comment]
So by solver you mean what is currently in Structural and Fluid (right?)

Say I write and FSI SOLVER, then I do exactly what you do above.

My concern is the stuff that goes to the Stage: Here you say BCs, Output,... etc is initialized. How am I supposed to know what the individual solvers need? Have a look at the current (old) Main Scripts. Even though Fluid and Structure are rather consistent now they still look very different!
This is knowledge that the App-Developers know, but I as user don't
Then it is not Black-Box any more

**Comment by RiccardoRossi** [comment]
by solver i intend fluid_solver, structure_solver, fsi_solver, composite_solver, thermal_solver ...etc etc...

BC processes belong to the stage not to the solver.

I don't understand in what sense they are "very different". They should expose all an identical API to the user

**Comment by philbucher** [comment]
They are different in the sense that simply the names in the ProjectParams are different (e.g. Structural has `constraints_processes` & `loads_processes` whereas fluid has `initial_conditions` etc...)
there are several cases of those, meaning that merging them breaks old cases.

Also if everything is in the BaseClass of the AnalysisStage, what will be be needing the ones in the Applications for?

**Comment by maceligueta** [comment]
Writing one single Json for couplings requires modifying the GUI a bit, while using several, independent Json files would allow developers to use the GUIs of the existing applications directly. I would not like to edit the Json manually each time I change the geometry (until the GUI for the coupling is created).

**Comment by philbucher** [comment]
+1 for @maceligueta 

I would really like to at least have the possibility do have one json for each "Physics"
Usually one has to debug each problem separately until it can be coupled and manually editing the json files is quite error prone!

**Pull request closed without merging**